Regression(251234@main): [Cairo][GTK][WPE] Darker output of SVG with filters fediffuselighting

pgorszkowski-igalia opened the pull request:
#### cfa8125ddf7c39a7256fcddf52b3c9f35de026e3
<pre>
Regression(251234@main): [Cairo][GTK][WPE] Darker output of SVG with filters fediffuselighting
<a href="https://bugs.webkit.org/show_bug.cgi?id=268537">https://bugs.webkit.org/show_bug.cgi?id=268537</a>

Reviewed by Said Abou-Hallawa and Nikolas Zimmermann.

Turn off the support for linearRGB color space in case of Cairo (251234@main) causes, that SVG
with filter, where linearRGB as inputs is demanded, generates darker output.

The SVG spec says that SourceGraphic has to be in linearRGB color space:
<a href="https://www.w3.org/TR/filter-effects-1/#attr-valuedef-in-sourcegraphic">https://www.w3.org/TR/filter-effects-1/#attr-valuedef-in-sourcegraphic</a>

In case of Cairo (which operates in SRGB color space) the image source (SourceGraphic)
should be created in SRGB and before passing it to filters, it should be transformed to
linearRGB color space.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
Remove some tests which pass with this change

* LayoutTests/svg/filters/feConvolveMatrix-clipped.svg:
* LayoutTests/svg/filters/feGaussianBlur-clipped.svg:
* LayoutTests/fast/gradients/conic-gradient-extended-stops.html:
Add possible differences in outputs

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::filteredNativeImage):
Pass color space to endDrawSourceImage

* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
Force to use SRGB as operating color space for DropShadow in case of Cairo.

* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
(WebCore::FilterImageTargetSwitcher::endClipAndDrawSourceImage):
(WebCore::FilterImageTargetSwitcher::endDrawSourceImage):
Before applying all filter, the source image has to be translated to requested
color space (only valid for Cairo).

* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h:
Added color space as additional parameter.

* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
(WebCore::FilterStyleTargetSwitcher::endDrawSourceImage):
WebCore::FilterStyleTargetSwitcher::endDrawSourceImage does not use color space so just ignore it

* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::applyFilterEffect):
Just pass DestinationColorSpace::SRGB() to endClipAndDrawSourceImage.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::applyResource):
In case of Cairo, the sourceImage has to be created with sRGB color space, because
Cairo operates in sRGB. The linearRGB color space will be passed to TargetSwitcher
in WebCore::LegacyRenderSVGResourceFilter::postApplyResource and the sourceImage will
be transformed to linearRGB before all filter applying.
(WebCore::LegacyRenderSVGResourceFilter::postApplyResource):
Before applying all filter, the source image has to be translated to requested
color space.

* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectsGraph):
Create SourceGraphic and SourceAlpha with LinearRGB color space if
color-interpolation_filter attribute of the Filter has LinearRGB value (only in case of Cairo).

Canonical link: <a href="https://commits.webkit.org/275091@main">https://commits.webkit.org/275091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c3f45b35def3924b8543d578fbf574fd916281a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40859 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36950 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22873 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17203 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33867 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41433 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16818 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14486 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14587 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36210 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44704 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34312 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37072 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36521 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40264 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40485 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15674 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38627 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17293 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35491 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47495 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9173 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17344 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9753 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16937 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WTF/wtf/PlatformEnable.h:
```diff
@@ -209,7 +209,7 @@
 
 #if !defined(ENABLE_DESTINATION_COLOR_SPACE_LINEAR_SRGB)
-#define ENABLE_DESTINATION_COLOR_SPACE_LINEAR_SRGB 0
+#define ENABLE_DESTINATION_COLOR_SPACE_LINEAR_SRGB 1
```
Maybe this can be enabled in the WPE/GTK cmake files instead of doing it for all platforms?

pgorszkowski-igalia replied:
By all platforms you mean everything except Cocoa (for it there is PlatformEnableCocoa.h), right?

philn replied:
yes

fujii replied:
If this is the right fix, we can enable it for all cairo ports. And if both Cococ and cairo ports enable ENABLE_DESTINATION_COLOR_SPACE_LINEAR_SRGB, we can remove the macro entirely.

pgorszkowski-igalia replied:
Ok, so I will remove ENABLE_DESTINATION_COLOR_SPACE_LINEAR_SRGB. I need also to check why css3/filters/* LayoutTests fail with my change.

nikolaszimmermann replied:
I agree with @fujii - we can remove it entirely. The fix looks correct -- for Cairo platforms color-space handling on ImageBuffers is a post-paint operation, Cairo provides sRGB data, if we want linearRGB, we need to convert after the pixels arrived in the ImageBuffer -- and endDrawSourceImage() is the right place for that.

+1 from my side.

pgorszkowski-igalia replied:
It will be removed in separate PR to make this change cleaner and increase readability of it.

nikolaszimmermann commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
```diff
@@ -41,7 +41,13 @@ FilterImageTargetSwitcher::FilterImageTargetSwitcher(GraphicsContext& destinatio
         return;
 
+#if USE(CAIRO)
+    UNUSED_PARAM(colorSpace);
```
I think it's unclear that that `colorSpace` parameter is now ignored for Cairo. We always enforce sRGB for the `m_sourceImage`, and auto-convert to linearRGB. While this is entirely correct for SVG filter processing -- it used to be possible to create a `FilterImageTargetSwitcher` operating entirely in sRGB, without doing any conversions (on non-Cairo platforms). With your patch that is no longer possible for Cairo, but still for Cocoa ports.

To avoid this different treatment between the Cocoa/Cairo ports, I suggest to store the `colorSpace` as `m_colorSpace` member in the class, and use it `endDrawSourceImage` to transform to `m_colorSpace` instead of hardcoding `DestinationColorSpace::LinearSRGB()`.

The net result of the patch stays identical, since SVG filter code passes LinearSRGB as colorSpace, it just removes a port discrepancy and surprises in the future if we ever add treatment for another color space.

pgorszkowski-igalia replied:
You are right, and that could be the reason of fails in case of css3/filters/* LayoutTests. Thanks Nikolas!!!

nikolaszimmermann reviewed (approved):
Looks nice, thanks @pgorszkowski-igalia. Are the css3 issues fixed? :-) EWS will tell us

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -308,6 +306,8 @@ RefPtr<FilterImage> SVGFilter::apply(FilterImage* sourceImage, FilterResults& re
 
             if (!sourceImage)
                 return nullptr;
+            else
```
No need for `else` since the `then` statement will return.

shallawa commented on Source/WebCore/platform/graphics/filters/FEDropShadow.h:
```diff
@@ -51,6 +51,10 @@ class FEDropShadow : public FilterEffect {
 
+#if !USE(CG)
+    void setOperatingColorSpace(const DestinationColorSpace&) override { }
+#endif
```
Why is this needed? Does `FEDropShadow` have to operate always in `SRGB()` on none CG ports?

pgorszkowski-igalia replied:
This change I added after analyzing the reason of the failing LayoutTests: 
* svg/filters/repeated-drop-shadow-effects.svg
* imported/blink/svg/filters/feDropShadow-linearrgb-flood-color.svg

```
<svg width="404" height="404" xmlns="http://www.w3.org/2000/svg">
    <defs>
        <filter id="repeated-drop-shadow" filterUnits="userSpaceOnUse"/>
    </defs>
    <rect x="200" y="200" width="4" height="4" style="fill:green; filter:url(#repeated-drop-shadow);"/>
    <script><![CDATA[
        let nameSpace = "http://www.w3.org/2000/svg";
        let filterElement = document.getElementById("repeated-drop-shadow");

        function createDropShadowEffects(count, dx, dy) {
            for (let i = 0; i < count; i++) {
                var dropShadowElement = document.createElementNS(nameSpace, "feDropShadow");
                dropShadowElement.setAttribute("dx", dx.toString());
                dropShadowElement.setAttribute("dy", dy.toString());
                dropShadowElement.setAttribute("stdDeviation", "0");
                dropShadowElement.setAttribute("flood-color", "green");
                filterElement.appendChild(dropShadowElement);
            }
        }

        createDropShadowEffects(50, -2,  2);
        createDropShadowEffects(50,  2, -2);
        createDropShadowEffects(50,  2,  2);
        createDropShadowEffects(50, -2, -2);
    ]]></script>
</svg>
```

So in this case we have (without this `void setOperatingColorSpace(const DestinationColorSpace&) override { }`):
* sourceImage is in linearRGB
* SourceGraphic is in linearRGB
* FEDropShadow operates in linearRGB but the `flood-color` is in sRGB(0,128,0,255) - green
* FEDropShadow output is in linearRGB (but with color values from the sRGB(0,128,0,255) - green)
* the final image should be in sRGB(deviceRGB) so the FEDropShadow output is transformed linearRGB->sRGB and the final image has pixels with values (0, 188, 0, 255) - brighten green

The green color in linearRGB(0, 55, 0, 255).
The green color in sRGB(0, 128, 0, 255).

So to solve this problem we have two options:
1. force DropShadow to operate in sRGB color space so the same as `flood-color` argument of this filer effect
2. transform `flood-color` (argument of the filter effect) to operating color space of FEDropShadow (which depends on color-interpolation-filter argument value - default `linearRGB`)

I chose option 1 but I think to be compliance with spec we should implement option 2.

@nikolaszimmermann , @shallawa : WDYT?

pgorszkowski-igalia replied:
I have checked how it is in Chromium and I found that `flood-color` is transformed to operating interpolation space: 
https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/blink/renderer/platform/graphics/filters/fe_drop_shadow.cc#76

pgorszkowski-igalia replied:
It seems that transforming a `shadow-color` for `DropShadow` filter is not a good option for Cairo because before the Color is passed to Cairo it is transformed to SRGB (so transformation in `FEDropShadowSoftwareApplier::apply` does not make sense):
https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/cairo/CairoUtilities.cpp#L118

So I propose a compromise and force SRGB operating color space in case of Cairo and DropShadow

shallawa commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
```diff
@@ -37,11 +37,19 @@ FilterImageTargetSwitcher::FilterImageTargetSwitcher(GraphicsContext& destinatio
+#else
     m_sourceImage = destinationContext.createScaledImageBuffer(m_sourceImageRect, filter.filterScale(), colorSpace, filter.renderingMode());
+#endif
+
```
Alternatively you can make `LegacyRenderSVGResourceFilter::applyResource()` pass `DestinationColorSpace::SRGB()` on  `USE(CAIRO)` when calling `FilterTargetSwitcher::create()`

pgorszkowski-igalia replied:
This solution was discussed with @nikolaszimmermann. I don't mind implementing your suggestion but would like to have @nikolaszimmermann opinion too.

nikolaszimmermann replied:
I have no strong opinion, both solutions are fine. @shallawa is the filter authority - if he prefers the latter solution, go for it ☺️

shallawa commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
```diff
@@ -79,6 +87,10 @@ void FilterImageTargetSwitcher::endDrawSourceImage(GraphicsContext& destinationC
+#if USE(CAIRO)
+    if (m_sourceImage)
+        m_sourceImage->transformToColorSpace(m_colorSpace);
+#endif
```
Instead of storing the `m_colorSpace` for `USE(CAIRO)`, you can add a new parameter `colorSpace` to this function. It will be passed as `DestinationColorSpace::LinearSRGB()` when it is called from `LegacyRenderSVGResourceFilter::postApplyResource()`.

pgorszkowski-igalia replied:
The same as in my answer in previous comment. @nikolaszimmermann I need your comment please:)

nikolaszimmermann replied:
Thanks @pgorszkowski-igalia - Saids suggestion is fine.

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -308,6 +306,8 @@ RefPtr<FilterImage> SVGFilter::apply(FilterImage* sourceImage, FilterResults& re
             if (!sourceImage)
                 return nullptr;
+            else
+                effect->setOperatingColorSpace(sourceImage->colorSpace());
```
I do not think this is the right place to set the `colorSpace` of the `SourceGraphic`. I think it should be passed to `SourceGraphic::create()` in `buildFilterEffectsGraph()` which should receive it as a new parameter from `SVGFilter::create()`. `LegacyRenderSVGResourceFilter::applyResource()` should pass `DestinationColorSpace::LinearSRGB()` to `SVGFilter::create()`.

shallawa replied:
But wait a second. Why we have to set the operating colorSpace of the `SourceGraphic` to the `colorSpace` of the `sourceImage`? The spec says
> The image is specified to be rendered in linear RGBA pixels.

And this is what we do when drawing the `sourceImage`. The ImageBuffer is created with `LinearRGB`. But the `SourceGraphic` can stay in `SRGB`. We will have to convert it only once.

In fact I think this change will introduce a pref regression when applying the SVGFilter. All the effects operate in `SRGB` unless the `effectElement.colorInterpolation()` says they should be in `LinearRGB`. So if you change the `SourceGraphic` to be `LinearRGB`, all the effects will have to convert the pixels of `SourceGraphic` to `SRGB` before operating on them.

pgorszkowski-igalia replied:
As you said the spec says:
> The image is specified to be rendered in linear RGBA pixels.

So the SourceGraphic (and SourceAlpha) should have operating color space depends on the `color-interpolation-filters` of the `Filter` element (which is by default linearRGB - spec: https://www.w3.org/TR/filter-effects-1/#propdef-color-interpolation-filters). Every filter effects can have its own `color-interpolation-filters` (but some of them ignore it, e.g: FEFlood (uses SRGB)).

So in case `<filter ...>` or `<filter color-interpolation-filters="linearRGB">` we should have:
* `sourceImage` in `linearRGB`
* SourceGraphic should operate in `linearRGB` so the output should be also in `linearRGB`

Example with `linearRGB`:
```
  <filter id="lightMe1">
    <feDiffuseLighting in="SourceGraphic" ...> or <feDiffuseLighting in="SourceGraphic" color-interpolation-filters="linearRGB" ...>...
```
* inputs of `feDiffuseLighting` should be transformed to linearRGB (SourceGraphic is in linearRGB so no-op)
* output of `feDiffuseLighting` should be also in linearRGB (also no-op)

Example with `sRGB`:
```
  <filter id="lightMe1">
    <feDiffuseLighting in="SourceGraphic" color-interpolation-filters="sRGB" ...>...
```
* inputs of `feDiffuseLighting` should be transformed to sRGB (SourceGraphic in this example is in linearRGB so it has to be transformed)
* output of `feDiffuseLighting` should be also in sRGB

pgorszkowski-igalia replied:
Anyway I will move this change (as @shallawa  suggested) to `buildFilterEffectsGraph()` and pass `DestinationColorSpace::LinearSRGB()` to `SourceGraphic::create()` if `filterElement.colorInterpolation() == ColorInterpolation::LinearRGB` (in other case will leave default value).

nikolaszimmermann replied:
It is indeed a complex topic. Would be great to get input from @shallawa / @heycam on the expected behavior.

nikolaszimmermann commented on Source/WebCore/platform/graphics/PlatformColorSpace.h:
```diff
@@ -46,9 +46,7 @@ class PlatformColorSpace {
 public:
     enum class Name : uint8_t {
         SRGB
-#if ENABLE(DESTINATION_COLOR_SPACE_LINEAR_SRGB)
```
It sounds like a good idea to remove this, but not in this PR, it is way too convoluted now, with those unimportant changes mixed with actual code changes. How about we remove the option in a follow-up PR once this has landed?

nikolaszimmermann commented on Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
```diff
@@ -65,20 +65,26 @@ void FilterImageTargetSwitcher::beginClipAndDrawSourceImage(GraphicsContext& des
         return;
 
     FilterResults results;
+#if USE(CAIRO)
```
I would appreciate a comment here to why this is different in CG and Cairo. (Skia will be more like CG here)

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -99,7 +99,8 @@ static std::optional<std::tuple<SVGFilterEffectsGraph, FilterEffectGeometryMap>>
 
-    SVGFilterEffectsGraph graph(SourceGraphic::create(), SourceAlpha::create());
+    SVGFilterEffectsGraph graph(filterElement.colorInterpolation() == ColorInterpolation::LinearRGB ? SourceGraphic::create(DestinationColorSpace::LinearSRGB()):SourceGraphic::create(),
+    filterElement.colorInterpolation() == ColorInterpolation::LinearRGB ? SourceAlpha::create(DestinationColorSpace::LinearSRGB()):SourceAlpha::create());
```
Maybe this is more readable:

```
auto colorSpace = filterElement.colorInterpolation() == ColorInterpolation::LinearRGB ? DestinationColorSpace::LinearSRGB() : DestinationColorSpace::SRGB();
SVGFilterEffectsGraph graph(SourceGraphic::create(colorSpace), SourceAlpha::create(colorSpace));
```

pgorszkowski-igalia committed:
Regression(251234@main): [Cairo][GTK][WPE] Darker output of SVG with filters fediffuselighting
https://bugs.webkit.org/show_bug.cgi?id=268537

Reviewed by Said Abou-Hallawa and Nikolas Zimmermann.

Turn off the support for linearRGB color space in case of Cairo (251234@main) causes, that SVG
with filter, where linearRGB as inputs is demanded, generates darker output.

The SVG spec says that SourceGraphic has to be in linearRGB color space:
https://www.w3.org/TR/filter-effects-1/#attr-valuedef-in-sourcegraphic

In case of Cairo (which operates in SRGB color space) the image source (SourceGraphic)
should be created in SRGB and before passing it to filters, it should be transformed to
linearRGB color space.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
Remove some tests which pass with this change

* LayoutTests/svg/filters/feConvolveMatrix-clipped.svg:
* LayoutTests/svg/filters/feGaussianBlur-clipped.svg:
* LayoutTests/fast/gradients/conic-gradient-extended-stops.html:
Add possible differences in outputs

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::filteredNativeImage):
Pass color space to endDrawSourceImage

* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
Force to use SRGB as operating color space for DropShadow in case of Cairo.

* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp:
(WebCore::FilterImageTargetSwitcher::endClipAndDrawSourceImage):
(WebCore::FilterImageTargetSwitcher::endDrawSourceImage):
Before applying all filter, the source image has to be translated to requested
color space (only valid for Cairo).

* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h:
Added color space as additional parameter.

* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
(WebCore::FilterStyleTargetSwitcher::endDrawSourceImage):
WebCore::FilterStyleTargetSwitcher::endDrawSourceImage does not use color space so just ignore it

* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::applyFilterEffect):
Just pass DestinationColorSpace::SRGB() to endClipAndDrawSourceImage.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::applyResource):
In case of Cairo, the sourceImage has to be created with sRGB color space, because
Cairo operates in sRGB. The linearRGB color space will be passed to TargetSwitcher
in WebCore::LegacyRenderSVGResourceFilter::postApplyResource and the sourceImage will
be transformed to linearRGB before all filter applying.
(WebCore::LegacyRenderSVGResourceFilter::postApplyResource):
Before applying all filter, the source image has to be translated to requested
color space.

* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectsGraph):
Create SourceGraphic and SourceAlpha with LinearRGB color space if
color-interpolation_filter attribute of the Filter has LinearRGB value (only in case of Cairo).

Canonical link: https://commits.webkit.org/275091@main
